pkg/deploy/internal/document/document.go

```diff
@@ -0,0 +1,167 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package document
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/clients/documents"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/idutils"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/entities"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/parameter"
+	deployErrors "github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/deploy/errors"
+	"github.com/go-logr/logr"
+)
+
+//go:generate mockgen -source=document.go -destination=document_mock.go -package=document documentClient
+type Client interface {
+	Get(ctx context.Context, id string) (documents.Response, error)
+	List(ctx context.Context, filter string) (documents.ListResponse, error)
+	Create(ctx context.Context, name string, externalId string, data []byte, documentType documents.DocumentType) (documents.Response, error)
+	Update(ctx context.Context, id string, name string, data []byte, documentType documents.DocumentType) (documents.Response, error)
+}
+
+var _ Client = (*DummyClient)(nil)
+
+type DummyClient struct{}
+
+// Create implements Client.
+func (c DummyClient) Create(ctx context.Context, name string, externalId string, data []byte, documentType documents.DocumentType) (documents.Response, error) {
+	return documents.Response{}, nil
+}
+
+// Get implements Client.
+func (c *DummyClient) Get(ctx context.Context, id string) (documents.Response, error) {
+	return documents.Response{}, nil
+}
+
+// List implements Client.
+func (c *DummyClient) List(ctx context.Context, filter string) (documents.ListResponse, error) {
+	return documents.ListResponse{}, nil
+}
+
+// Update implements Client.
+func (c *DummyClient) Update(ctx context.Context, id string, name string, data []byte, documentType documents.DocumentType) (documents.Response, error) {
+	return documents.Response{}, nil
+}
+
+func Deploy(ctx context.Context, client Client, properties parameter.Properties, renderedConfig string, c *config.Config) (entities.ResolvedEntity, error) {
+	// create new context to carry logger
+	ctx = logr.NewContext(ctx, log.WithCtxFields(ctx).GetLogr())
+
+	documentName, ok := properties[config.NameParameter].(string)
+	if !ok {
+		return entities.ResolvedEntity{}, errors.New("missing name parameter")
+	}
+
+	documentType, err := getDocumentTypeFromConfigType(c.Type)
+	if err != nil {
+		return entities.ResolvedEntity{}, fmt.Errorf("cannot get document type for config: %w", err)
+	}
+
+	// strategy 1: if an origin id is available, try to update that document
```

Comment:
What do you think of putting this different strategy into different functions?

Reply:
I'll give it a try to see if it makes it any clearer

Reply:
What do you think of the next order:
 * in case `ObjectOriginId` exists, use it (`id := c.ObjectOriginId`). Otherwise, try to find id via `externalId`
 * in case `id` is known, update the existing; otherwise, create a new entry.

With this approach, there will be only one place where the `client.Update` is called.

Reply:
Two updates are indeed required as from the spec we must try with `originObjectId` and fall back to using `externalId` if the update doesn't work

Reply:
So, when the `originObjectID` is present, but there isn't a document, we try to find a document with the matching `externalId`?
What in the case where on the server exists both: a document with the same 'originObjectID` and another one that matches `externalId`?

Reply:
Correct, if `originObjectId` is present and there is a document with that ID then we use that. It has precedence over `externalId`.